components/script/dom/textdecodercommon.rs

```diff
@@ -0,0 +1,136 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::{Cell, RefCell};
+
+use encoding_rs::{Decoder, DecoderResult, Encoding};
+use script_bindings::str::DOMString;
+
+use crate::dom::bindings::codegen::UnionTypes::ArrayBufferViewOrArrayBuffer;
+use crate::dom::bindings::error::{Error, Fallible};
+
+/// The shared part of `TextDecoder` and `TextDecoderStream`
+///
+/// Note that this does NOT correspond to the `TextDecoderCommon`
+/// interface defined in the WebIDL
+#[allow(non_snake_case)]
+#[derive(JSTraceable, MallocSizeOf)]
+pub(crate) struct TextDecoderCommon {
+    #[no_trace]
+    encoding: &'static Encoding,
+    fatal: bool,
+    ignoreBOM: bool,
+    #[ignore_malloc_size_of = "defined in encoding_rs"]
+    #[no_trace]
+    decoder: RefCell<Decoder>,
+    in_stream: RefCell<Vec<u8>>,
+    do_not_flush: Cell<bool>,
+}
+
+#[allow(non_snake_case)]
+impl TextDecoderCommon {
+    pub(crate) fn new_inherited(
+        encoding: &'static Encoding,
+        fatal: bool,
+        ignoreBOM: bool,
+    ) -> TextDecoderCommon {
+        let decoder = if ignoreBOM {
+            encoding.new_decoder()
+        } else {
+            encoding.new_decoder_without_bom_handling()
+        };
+
+        TextDecoderCommon {
+            encoding,
+            fatal,
+            ignoreBOM,
+            decoder: RefCell::new(decoder),
+            in_stream: RefCell::new(Vec::new()),
+            do_not_flush: Cell::new(false),
+        }
+    }
+
+    pub(crate) fn encoding(&self) -> DOMString {
+        DOMString::from(self.encoding.name().to_ascii_lowercase())
+    }
+
+    pub(crate) fn fatal(&self) -> bool {
+        self.fatal
+    }
+
+    pub(crate) fn ignore_bom(&self) -> bool {
+        self.ignoreBOM
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) fn decode(
+        &self,
+        input: Option<&ArrayBufferViewOrArrayBuffer>,
+        do_not_flush: bool,
+    ) -> Fallible<String> {
+        // Step 1.
```

Comment:
since you are touching this code, can you write the correct spec steps and avoid numbers